File: king_phisher/sms.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  king_phisher/sms.py
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following disclaimer
#    in the documentation and/or other materials provided with the
#    distribution.
#  * Neither the name of the project nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

import argparse
import random
import re
import smtplib
import sys

import dns.resolver
import smoke_zephyr.utilities

if sys.version_info[0] < 3:
	from email.MIMEText import MIMEText
else:
	from email.mime.text import MIMEText

__version__ = '1.0'
__all__ = ('lookup_carrier_gateway', 'send_sms')

CARRIERS = {
	'AT&T':          'txt.att.net',
	'Boost':         'myboostmobile.com',
	'Sprint':        'messaging.sprintpcs.com',
	'T-Mobile':      'tmomail.net',
	'Verizon':       'vtext.com',
	'Virgin Mobile': 'vmobl.com',
}
"""A dictionary for mapping carrier names to SMS via email gateways."""

DEFAULT_FROM_ADDRESS = 'sms@king-phisher.com'
"""The default email address to use in the from field."""

@smoke_zephyr.utilities.Cache('6h')
def get_smtp_servers(domain):
	"""
	Get the SMTP servers for the specified domain by querying their MX records.

	:param str domain: The domain to look up the MX records for.
	:return: The smtp servers for the specified domain.
	:rtype: list
	"""
	mx_records = dns.resolver.query(domain, 'MX')
	return [str(r.exchange).rstrip('.') for r in mx_records]

def normalize_name(name):
	return name.lower().replace('&', '').replace('-', '')

def lookup_carrier_gateway(carrier):
	"""
	Lookup the SMS gateway for the specified carrier. Normalization on the
	carrier name does take place and if an invalid or unknown value is
	specified, None will be returned.

	:param str carrier: The name of the carrier to lookup.
	:return: The SMS gateway for the specified carrier.
	:rtype: str
	"""
	carrier = normalize_name(carrier)
	carrier_address = [c for c in CARRIERS.keys() if normalize_name(c) == carrier]
	if len(carrier_address) != 1:
		return None
	return CARRIERS[carrier_address[0]]

def send_sms(message_text, phone_number, carrier, from_address=None):
	"""
	Send an SMS message by emailing the carriers SMS gateway. This method
	requires no money however some networks are blocked by the carriers
	due to being flagged for spam which can cause issues.

	:param str message_text: The message to send.
	:param str phone_number: The phone number to send the SMS to.
	:param str carrier: The cellular carrier that the phone number belongs to.
	:param str from_address: The optional address to display in the 'from' field of the SMS.
	:return: This returns the status of the sent message.
	:rtype: bool
	"""
	from_address = (from_address or DEFAULT_FROM_ADDRESS)
	phone_number = phone_number.replace('-', '').replace(' ', '')
	# remove the country code for these 10-digit based
	match = re.match('1?(?P<phone_number>[0-9]{10})', phone_number)
	if match is None:
		raise ValueError('the phone number appears invalid')
	phone_number = match.group('phone_number')

	if len(message_text) > 160:
		raise ValueError('message length exceeds 160 characters')
	message = MIMEText(message_text)

	carrier_address = lookup_carrier_gateway(carrier)
	if not carrier_address:
		raise ValueError('unknown carrier specified')

	to_address = "{0}@{1}".format(phone_number, carrier_address)
	message['To'] = to_address
	message['From'] = from_address

	sms_gateways = get_smtp_servers(carrier_address)
	random.shuffle(sms_gateways)
	message_sent = False
	for sms_gateway in sms_gateways:
		try:
			smtp_connection = smtplib.SMTP(sms_gateway)
			smtp_connection.sendmail(from_address, [to_address], message.as_string())
			smtp_connection.quit()
		except (smtplib.SMTPConnectError, smtplib.SMTPDataError, smtplib.SMTPHeloError):
			continue
		message_sent = True
		break
	return message_sent

def _argp_sms_carrier_type(arg):
	if not lookup_carrier_gateway(arg):
		raise argparse.ArgumentTypeError("{0} is not a valid sms carrier".format(repr(arg)))
	return arg

def main():
	parser = argparse.ArgumentParser(description='Send SMS Messages', conflict_handler='resolve')
	parser.add_argument('--from', dest='from_email', default=DEFAULT_FROM_ADDRESS, help='the email address to send the message from')
	parser.add_argument('-v', '--version', action='version', version=parser.prog + ' Version: ' + __version__)
	parser.add_argument('phone_number', help='destination phone number')
	parser.add_argument('carrier', type=_argp_sms_carrier_type, help='target carrier')
	parser.add_argument('message', help='message text to send')
	results = parser.parse_args()

	if send_sms(results.message, results.phone_number, results.carrier, from_address=results.from_email):
		print('[+] Successfully Sent!')  # pylint: disable=superfluous-parens
	else:
		print('[-] Failed To Send')  # pylint: disable=superfluous-parens
	return 0

if __name__ == '__main__':
	main()

File: data/msf/sms.rb
# Copyright (c) 2015, Brandan [coldfusion]
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following disclaimer
#    in the documentation and/or other materials provided with the
#    distribution.
#  * Neither the name of the  nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


require 'uri'

module Msf
  class Plugin::SessionSMS < Msf::Plugin
    include Msf::SessionEvent

    CARRIERS = ['AT&T', 'Boost', 'Sprint', 'T-Mobile', 'Verizon', 'Virgin Mobile']
    private_constant :CARRIERS

    def initialize(framework, opts)
      super
      add_console_dispatcher(SMSCommandDispatcher)
    end

    def cleanup
      self.framework.events.remove_session_subscriber(@inst)
      remove_console_dispatcher('sms')
    end

    def name
      "sms"
    end

    def desc
      "Sends a SMS message when recieving a new session through the use of the King Phisher's REST API."
    end

    def sms_yaml
      "#{Msf::Config.get_config_root}/sms.yaml"
    end

    class SMSCommandDispatcher < Plugin::SessionSMS
      include Msf::Ui::Console::CommandDispatcher

      def initialize(framework)
        super
        @king_phisher_server = nil
        @king_phisher_token = nil
        @sms_number = nil
        @sms_carrier = nil

        print_status('Loaded previously configured settings') if read_settings
      end

      def on_session_open(session)
        print_status('Session received, sending SMS...')
        send_sms("Session: #{session.sid} opened")
      end

      def name
        'sms'
      end

      def read_settings
        return false unless File.exist?(sms_yaml)

        ldconfig = YAML.load_file(sms_yaml)
        @king_phisher_server = ldconfig['king_phisher_server']
        @king_phisher_token = ldconfig['king_phisher_token']
        @sms_number = ldconfig['sms_number']
        @sms_carrier = ldconfig['sms_carrier']
        return true
      end

      def send_sms(message)
        message = Rex::Text.uri_encode(message)
        begin
          http_client = Rex::Proto::Http::Client.new(@king_phisher_server)
          http_client.connect
          request = http_client.request_cgi({
            'uri'    => '/_/api/sms/send',
            'query'  => "token=#{@king_phisher_token}&message=#{message}&phone_number=#{@sms_number}&carrier=#{@sms_carrier}"
          })
          res = http_client.send_recv(request)
        rescue Exception => e
          print_error("Error sending SMS: #{e.class} - #{e}")
          return false
        ensure
          http_client.close
        end

        res and res.code == 200
      end

      def commands
        {
          'sms_start'       => 'Start SMS alerts for new sessions',
          'sms_stop'        => 'Stop SMS alerts for new sessions',
          'sms_save'        => "Save SMS settings to #{sms_yaml}",
          'sms_set_server'  => 'Set domain name of the King Phisher server',
          'sms_set_token'   => 'Set King Phisher\'s API token',
          'sms_set_number'  => 'Set number to send SMS alerts to on new session',
          'sms_set_carrier' => 'Set carrier for sending SMS messages',
          'sms_show_params' => 'Shows currently set or saved parameters',
          'sms_test'        => 'Send a test SMS message to verify the parameters'
        }
      end

      def cmd_sms_start
        unless @king_phisher_server and @king_phisher_token and @sms_number and @sms_carrier
          print_error('You have not provided all the parameters!')
          return
        end

        self.framework.events.add_session_subscriber(self)
        print_good('Started SMS sessions session notifications')
      end

      def cmd_sms_stop
        self.framework.events.remove_session_subscriber(self)
        print_good('Stopped SMS sessions session notifications')
      end

      def cmd_sms_save
        unless @king_phisher_server and @king_phisher_token and @sms_number and @sms_carrier
          print_error('You have not provided all the parameters!')
          return
        end

        config = {
          'king_phisher_server' => @king_phisher_server,
          'king_phisher_token'  => @king_phisher_token,
          'sms_number'          => @sms_number,
          'sms_carrier'         => @sms_carrier
        }
        File.open(sms_yaml, 'w') do |out|
          YAML.dump(config, out)
        end
        print_good("All parameters saved to #{sms_yaml}")
      end

      def cmd_sms_test(*args)
        unless @king_phisher_server and @king_phisher_token and @sms_number and @sms_carrier
          print_error('You have not provided all the parameters!')
          return
        end

        message = args.shift
        message = 'Test SMS message' if message.nil?

        if send_sms(message)
          print_good('Sent the test SMS message')
        else
          print_error('Failed to send the test SMS message')
        end
      end

      def cmd_sms_set_server(*args)
        if args.length > 0
          print_status("Setting the King Phisher server to #{args[0]}")
          @king_phisher_server = args[0]
        else
          print_error('Please provide the domain name of your King Phisher server!')
        end
      end

      def cmd_sms_set_token(*args)
        if args.length > 0
          print_status("Setting King Phisher's REST API token to #{args[0]}")
          @king_phisher_token = args[0]
        else
          print_error('Please provide the REST API token of your King Phisher server!')
        end
      end

      def cmd_sms_set_number(*args)
        if args[0].length == 10
          print_status("Setting SMS number to #{args[0]}")
          @sms_number = args[0]
        else
          print_error('Please provide a valid SMS number!')
        end
      end

      def cmd_sms_set_carrier(*args)
        if args.length > 0 and CARRIERS.include?(args[0])
          print_status("Setting SMS carrier to #{args[0]}")
          @sms_carrier = args[0]
        else
          print_error("Please provide a valid SMS carrier (#{CARRIERS.join(', ')})!")
        end
      end

      def cmd_sms_show_params
        print_status('Parameters:')
        print_status("  King Phisher Server: #{@king_phisher_server}")
        print_status("  King Phisher Token: #{@king_phisher_token}")
        print_status("  SMS Number: #{@sms_number}")
        print_status("  SMS Carrier: #{@sms_carrier}")
      end
    end
  end
end
